### DISCUSSION TITLE
Why Intune ESP check removed

### DISCUSSION THREAD

**Discussion by Pollewops** [PSAppDeployToolkit/PSAppDeployToolkit#1067] [General]
PSADT 3.9.3 had a check for ESP, and when detected run de application silent.

Why this check is removed in 3.10.x versions?

**Comment by DanGough** [comment]
It was changed rather than removed - if it is not working as expected then we can file this as a bug.

v3.9.3:
```powershell
# Check if user session is running under defaultuser0 account (Autopilot OOBE) or if application is installing during ESP and if so change deployment to run silently
If ($RunAsActiveUser.NTAccount -like '*\defaultuser0' -and (((Get-Process -Name 'wwahost' -ErrorAction 'SilentlyContinue').count) -gt 0)) {
    Write-Log -Message "Autopilot OOBE user [$($CurrentLoggedOnUserSession.UserName)] or ESP process 'wwahost' detected, changing deployment mode to silent." -Source $appDeployToolkitExtName
    $deployMode = 'Silent'
}
```

v3.10.2:
```powershell
# Check if OOBE / ESP is running [credit Michael Niehaus]
$TypeDef = @"
using System;
using System.Text;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Api
{
 public class Kernel32
 {
   [DllImport("kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
   public static extern int OOBEComplete(ref int bIsOOBEComplete);
 }
}
"@

Add-Type -TypeDefinition $TypeDef -Language CSharp

$IsOOBEComplete = $false
$hr = [Api.Kernel32]::OOBEComplete([ref] $IsOOBEComplete)

If (!($IsOOBEComplete)) {
    Write-Log -Message "Detected OOBE in progress, changing deployment mode to silent." -Source $appDeployToolkitName
    $deployMode = 'Silent'
}
```

**Comment by Pollewops** [comment]
Hi Dan,I use 3.10.1 version and in there it is not working.I now manually add couple of lines myself to check for ESP. Let you know tomorrow if that is working.Met een glimlach,JohanOp 24 okt 2024 om 11:08 heeft Dan Gough ***@***.***> het volgende geschreven:﻿
It was changed rather than removed - if it is not working as expected then we can file this as a bug.
v3.9.3:
# Check if user session is running under defaultuser0 account (Autopilot OOBE) or if application is installing during ESP and if so change deployment to run silently
If ($RunAsActiveUser.NTAccount -like '*\defaultuser0' -and (((Get-Process -Name 'wwahost' -ErrorAction 'SilentlyContinue').count) -gt 0)) {
    Write-Log -Message "Autopilot OOBE user [$($CurrentLoggedOnUserSession.UserName)] or ESP process 'wwahost' detected, changing deployment mode to silent." -Source $appDeployToolkitExtName
    $deployMode = 'Silent'
}
v3.10.2:
# Check if OOBE / ESP is running [credit Michael Niehaus]
$TypeDef = @"
using System;
using System.Text;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace Api
{
 public class Kernel32
 {
   [DllImport("kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
   public static extern int OOBEComplete(ref int bIsOOBEComplete);
 }
}
"@

Add-Type -TypeDefinition $TypeDef -Language CSharp

$IsOOBEComplete = $false
$hr = [Api.Kernel32]::OOBEComplete([ref] $IsOOBEComplete)

If (!($IsOOBEComplete)) {
    Write-Log -Message "Detected OOBE in progress, changing deployment mode to silent." -Source $appDeployToolkitName
    $deployMode = 'Silent'
}

**Comment by Pollewops** [comment]
Ok I see 3.10.1 does also have this ESP function included, which seems not to work for me. I am now checking with manually added these lines:

    # Verify if started during ESP
    $ESPCloudHost = @(Get-CimInstance -ClassName 'Win32_Process' -Filter "Name like 'CloudExperienceHostBroker.exe'" -ErrorAction 'SilentlyContinue')
    $ESPWWAHost = @(Get-CimInstance -ClassName 'Win32_Process' -Filter "Name like 'WWAHost.exe'" -ErrorAction 'SilentlyContinue')
    $ESPCount = 0
    $ESPCount = $ESPCloudHost.Count + $ESPWWAHost.Count
    If ($ESPCount -ne 0)  { $DeployMode = 'Silent' }

**Comment by mjr4077au** [comment]
The way we're testing whether the ESP is running via that P/Invoke is the correct way to do this and has been tested/proven time and time again in deployments at my work. Please test all this again on v4 when it's released and see whether you're having any further issues.

**Reply by mjr4077au** [reply]
To circle back on this discussion, as mentioned in https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/issues/1072, there is potentially an issue with our `OOBEComplete()` P/Invoke in v3.x that has been properly resolved in v4.
